service/header/sync.go

```diff
@@ -330,3 +427,120 @@ func (s *Syncer) getHeaders(ctx context.Context, start, amount uint64) ([]*Exten
 
 	return out, nil
 }
+
+// heightSub provides a way to wait until a specific height becomes available and synced
+type heightSub struct {
+	// usually we don't attach context to structs, but this is an exception
+	// as GetByHeight and ProvideHeights needs to access it
+	ctx       context.Context
+	cancel    context.CancelFunc
+	store     Store
+	height    uint64
+	provideCh chan []*ExtendedHeader
+	reqsCh    chan *heightReq
+	reqs      map[uint64][]chan *heightResp
+}
+
+type heightReq struct {
+	resp   chan *heightResp
+	height uint64
+}
+
+type heightResp struct {
+	header *ExtendedHeader
+	err    error
+}
+
+func newHeightSub(store Store) *heightSub {
+	return &heightSub{
+		store:     store,
+		provideCh: make(chan []*ExtendedHeader, 4),
+		reqsCh:    make(chan *heightReq, 4),
+		reqs:      make(map[uint64][]chan *heightResp),
```

Comment:
why can there be multiple heightResps per height request?

Reply:
A similar case to: "why there can be multiple header.Subscriptions?". There might be an RPC call to get a Header that can wait until it is available, so not just DAS would use it. Similarly, at some point, we might develop some other component that may wait for a header. Also, anyone who accesses the embedded Node in their app may want to access this API to wait for a Header. And finally, @Bidon15 wanted to wait for a specific height. All those cases will access the same heightSub instance at the same time and can potentially subscribe over the same height, which is why we need to handle multiple subscriptions.

Reply:
ack